torch/_decomp/decompositions.py

```diff
@@ -1840,6 +1840,111 @@ def _native_batch_norm_legit_functional(
     return output, save_mean, save_rstd, new_running_mean, new_running_var
 
 
+def _get_batch_norm_reserve_tensor(
+    input: Tensor,
+    weight: Optional[Tensor],
+    bias: Optional[Tensor],
+    running_mean: Tensor,
+    running_var: Tensor,
+    eps: float,
+    cudnn_enabled: bool,
+) -> Tensor:
+    backend = torch._C._select_batch_norm_backend(  # type: ignore[attr-defined]
+        input, weight, bias, running_mean, running_var, True, eps, cudnn_enabled
+    )
+    if backend == torch._C._BatchNormBackend.Cudnn:  # type: ignore[attr-defined]
+        reserve_size = torch._C._get_cudnn_batch_norm_reserve_space_size(input)  # type: ignore[attr-defined]
+        return torch.empty(
+            reserve_size, dtype=input.dtype, layout=input.layout, device=input.device
+        )
+    else:
+        return Tensor()
+
+
+@register_decomposition(aten.batch_norm_with_update.default)
+def batch_norm_with_update(
+    input: Tensor,
+    weight: Optional[Tensor],
+    bias: Optional[Tensor],
+    running_mean: Tensor,
+    running_var: Tensor,
+    momentum: float,
+    eps: float,
+    cudnn_enabled: bool,
+) -> Tuple[Tensor, Tensor, Tensor, Tensor]:
+    output, save_mean, save_rstd, _, _ = native_batch_norm_helper(
+        input,
+        weight,
+        bias,
+        running_mean,
+        running_var,
+        True,  # training
+        momentum,
+        eps,
+        False,  # functional
+    )
+    reserve = _get_batch_norm_reserve_tensor(
```

Comment:
I think the more contentious part of this PR is the handling for the extra tensor that cudnn saves for backward, so I was hoping to see what @albanD thinks of this approach.

Right now in this PR, the new batch norm op unconditionally returns an extra tensor, and gives it the shape that cudnn needs if we detect that we would have run cudnn in eager mode, and otherwise returns a size-0 tensor.

This works around the problem of `native_batch_norm` and `cudnn_batch_norm` returning different arguments in eager mode (which the legit op did not address).

It also relies on AOTAutograd DCE'ing this tensor when we trace out the forward/joint graph, to make sure we do not regress perf by forcing inductor to allocate this tensor (which... I think should work, but we should test).

Reply:
Since we specialize on if cudnn is enabled or not, I think this is fine to do.
As we discussed, if inductor doesn't want to see cudnn ops and specificities, we should just disable it when compiling for that particular backend. 

cc @jansel do you have an opinion on this?